materials_io/base.py

```diff
@@ -11,88 +13,91 @@ class BaseParser(ABC):
-            else:
-                # If is_valid, is base implementation, run fill parser anyway
-                try:
-                    yield group, self.parse(group, context)
-                except Exception:
-                    continue
+        # Iterate over all directories
+        for root, dirs, files in os.walk(path):
+            # Generate the full paths for all of the contents
+            full_paths = map(lambda x: os.path.join(root, x), files + dirs)
+
+            # Parse each identified group
+            for group in self.group(full_paths, context):
+                yield (group, self.parse(group, context))
 
     @abstractmethod
-    def parse(self, group, context=None):
+    def parse(self, group: List[str], context: dict = None) -> dict:
         """Extract metadata from a group of files
 
+        A group of files is a set of 1 or more files that describe the same object, and will be
+        be used together to create s single summary record.
+
         Arguments:
-            group (list of str):  A list of one or more files to parse as a unit.
-            context (dict): An optional data context/configuration dictionary. Default None.
+            group (list of str):  A list of one or more files that should be parsed together
+            context (dict): Context about the files
 
         Returns:
-            (list of dict): The parsed results, in JSON-serializable format.
+            (dict): The parsed results, in JSON-serializable format.
         """
-        pass
 
-    def is_valid(self, group, context=None):
-        """Determine whether a group of files is compatible with this parser
+    def parse_as_unit(self, files: List[str]) -> dict:
+        """Parse a group of files and merge their metadata
 
-        Arguments:
-            group (list of str):  A list of one or more files to parse as a unit.
-            context (dict): An optional data context/configuration dictionary. Default None.
+        Used if each file in a group are parsed separately, but the resultant metadata
+        should be combined after parsing.
 
+        Args:
+            files ([str]): List of files to parse
         Returns:
-            (bool): Whether the group can be parsed by the parser
+            (dict): Metadata summary from
         """
-        try:
-            res = self.parse(group, context=context)
-            if not res:
-                raise ValueError
-        except Exception:
-            return False
-        else:
-            return True
-
-    def group(self, root, context=None):
-        """Identify sets files in a directory that are related to each other
+
+        # Initialize output dictionary
+        metadata = {}
+
+        # Loop over all files
+        for group in self.group(files):
+            try:
+                record = self.parse(group)
+                print(record)
+            except Exception as exc:
+                raise exc
+                continue
```

Comment:
It looks like this might need to be cleaned up. Were the re-raise and print statements for debugging?

Reply:
Oops! That was for debugging